libdrgn/program.c

```diff
@@ -1188,6 +1188,24 @@ drgn_program_find_thread_userspace_live(struct drgn_program *prog, uint32_t tid,
 		(*ret)->prog = prog;
 		(*ret)->tid = tid;
 		(*ret)->prstatus = (struct nstring){};
+
+		const int path_len = strlen(path);
+		const int comm_file_path_len = path_len + 6;
+		char *comm_file_path = malloc(comm_file_path_len);
+
+		if (!comm_file_path)
+			return &drgn_enomem;
+		strncpy(comm_file_path, path, path_len);
+		strncat(comm_file_path, "/comm", 5);
+		strncat(comm_file_path, "\0", 1);
```

Comment:
We can avoid the extra allocation and all of this string processing if we reserve enough space in `path` to append `"/comm"`:

```diff
diff --git a/libdrgn/program.c b/libdrgn/program.c
index 1a5103e..83b5a21 100644
--- a/libdrgn/program.c
+++ b/libdrgn/program.c
@@ -1177,7 +1177,7 @@ drgn_program_find_thread_userspace_live(struct drgn_program *prog, uint32_t tid,
 		- sizeof("%ld%" PRIu32)
 		+ max_decimal_length(long)
 		+ max_decimal_length(uint32_t)
-		+ 1];
+		+ sizeof("/comm")];
 	snprintf(path, sizeof(path), FORMAT, (long)prog->pid, tid);
 #undef FORMAT
 	int r = access(path, F_OK);
@@ -1188,6 +1188,11 @@ drgn_program_find_thread_userspace_live(struct drgn_program *prog, uint32_t tid,
 		(*ret)->prog = prog;
 		(*ret)->tid = tid;
 		(*ret)->prstatus = (struct nstring){};
+
+		strcat(path, "/comm");
+		FILE *comm_file = fopen(path, "r");
+		...
+
 		return NULL;
 	} else if (errno == ENOENT) {
 		*ret = NULL;
```

(Note that `sizeof("/comm")` includes the null terminator, and `strcat()` copies the null terminator).